feat: add Tabs component

joshuatai opened the pull request.

Commit:
feat: add Tabs component

Commit:
feat: add Tabs component

Commit:
feat: add Tabs component

Commit:
Merge branch 'dev' into tabs

Commit:
fix: remove unuse Tabs functions - size, custom tab, enclosed-colored, soft-rounded, and solid-rounded tabs

Commit:
fix: remove unuse Tabs functions - size, custom tab, enclosed-colored, soft-rounded, and solid-rounded tabs

joshuatai committed:
Merge branch 'dev' into feature/tabs-component

Commit:
Merge branch 'feature/tabs-component' of github.com:trendmicro-frontend/styled-ui into feature/tabs-component

tinaClin commented on packages/react-styled-ui/src/Tabs/index.js:
```diff
@@ -0,0 +1,8 @@
+import TabPanel from './TabPanel';
+
+export default Tabs;
+export { TabList, Tab, TabPanel, TabPanels };
```
要寫成
`export { Tabs, TabList, Tab, TabPanel, TabPanels };`

不然`Tabs`會拿不到。
`import { Tabs, TabList, Tab, TabPanel, TabPanels } from 'app/components/Tabs';`

joshuatai replied:
好的喔～～已改～～

tinaClin commented on packages/react-styled-ui/src/Tabs/TabList.js:
```diff
@@ -0,0 +1,116 @@
+    const handleClick = (event) => {
+      // Hack for Safari. Buttons don't receive focus on click on Safari
+      // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+      allNodes.current[index].focus();
```
我在實作時，有發生`allNodes.current[index]`是`undefined`的情況。

joshuatai replied:
這個我沒辦法 reproduce 這個 error，研判應該是單獨拿 Tabs 的 source code 和 project 一起 build 的因素。如果是和整個 styled ui 一起 build 出來的應該就不會。我有把 build 好的拉進 local projct 裡使用，沒有發生這個 error。 Apply themeProvider 和 colorModeProvider 也沒問題喔～～～
![image](https://user-images.githubusercontent.com/1780412/84659129-8100ad00-af49-11ea-987d-7005d7d57b71.png)

![image](https://user-images.githubusercontent.com/1780412/84659001-54e52c00-af49-11ea-9737-b23c451a1c7d.png)

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+  },
+};
+
+export const baseProps = {
```
有沒有考慮把`tab`和`tabList`的props分開寫會比較清楚?

joshuatai replied:
查客拉可能因為有再 export baseProps，所以才寫在一個 baseProps 裡，然後使用 .tab 或 .tabList 拿，不過如果我們沒有 export 的需求，的確可以寫分開～～～我把它改掉

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+  },
+  tabList: {
+    borderWidth: 0,
+    borderBottomWidth: '2px',
```
為什麼`tabList`要有`borderBottomWidth: '2px',`?

joshuatai replied:
因為那是為了留給萬一有底線的需求～～ Pin 說也要有預留萬一有需要這條線的狀況。其實查克拉也有這條～～

![image](https://user-images.githubusercontent.com/1780412/84725543-7122b080-afbd-11ea-8968-87a0f473ca43.png)

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+  }
+};
+
+export const statusProps = {
```
我個人覺得不用`statusProps`，直接寫在`Pseudo-classes`裡。

joshuatai replied:
我覺得還是放在 styles 裡面比較好，因為它 useTabStyle 裡的實作概念就是不同排列組合做合成，而且是有優先順序的。如果只有 status 抽出去， 從 styles 的 code 很難發現有一個 status 的 style 寫在 pseudo 上，而且它永遠會 overwrite 所有的 style 設定，所以如果 styles 不管怎麼改動都沒用，反而要去找為什麼被 overwrite 了。
![image](https://user-images.githubusercontent.com/1780412/84737429-82c68100-afda-11ea-8369-54eb1e09e56a.png)

joshuatai replied:
Merge statusProps into each variant style instead of using the spread object to overwrite properties

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+  let _focusBorderColor = 'blue:60';
+  let _selectedBg = 'gray:80';
+  let _selectedBorder = _selectedBg;
+  if (colorMode === 'light') {
```
希望這裡不用`if`去判斷`drak`和`light`，用object mapping

joshuatai replied:
Got it~~ 哪根筋不對～～上面寫 object map，下面寫 condition 判斷 XD

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+        borderWidth: _focusBorderWidth,
+        px: `calc(${_px} - ${_focusBorderWidth})`,
+        py: `calc((${theme.space[tabSizes[size].height]} - ${_lineHeight} - (${_focusBorderWidth} * 2)) / 2)`,
+        zIndex: 3
```
像這樣子，直接寫在這裡，就不用追`statusProps`了。

joshuatai replied:
@tinaClin , 不建議這樣做，因為 statusPros 在 useTabStyle 裡是有權重的，查克拉把它拉進 useTabsStyle 裡是有用意的。拉出去外面的好處是什麼呢？

roth1002 replied:
library以用的人好用為目的   statusProps就算拿出去外面他也是要追

更何況 styleProps通常是有權重的XDD

tinaClin replied:
我想說，在`statusProps`已經設定
```
_focus: {
   zIndex: '3'
},
```
然後在這裡又寫一次 `zIndex: 3`，

既然如此為什麼不直接在return object直接寫zIndex，還要多一個statusProps？

joshuatai replied:
Merge statusProps into each variant style instead of using the spread object to overwrite properties

tinaClin commented on packages/react-styled-ui/src/Tabs/styles.js:
```diff
@@ -0,0 +1,264 @@
+        borderColor
+      })
+    },
+    // tabList: {
```
用不到的code，直接刪掉吧~

joshuatai replied:
XD

Commit:
fix: export Tabs for individual import

cheton committed:
Merge branch 'dev' into feature/tabs-component

Commit:
fix: - use object map to replace condition of color mode
     - export tabs component gather with Tab, TabList and other components
     - separate tab and tabList baseProps style
     - remove unuse code

Commit:
fix: add a space before }

Commit:
Merge branch 'feature/tabs-component' of github.com:trendmicro-frontend/styled-ui into feature/tabs-component

Commit:
fix: remove duplicated Tabs import

Commit:
fix: merge status style into each variant style instead of using spread operator

cheton commented on packages/styled-ui-docs/pages/tabs.mdx:
```diff
@@ -0,0 +1,282 @@
+| `onChange`     | `(index: number) => void`                                                         |              | The callback to update the active tab index.                                                                                    |
+| `index`        | `number`                                                                          |              | The controlled index of the tabs.                                                                                               |
+| `defaultIndex` | `number`                                                                          |              | The index of the initial active tab.                                                                                            |
+| `isManual`     | `boolean`                                                                         |              | If `true`, keyboard navigation changes focus between tabs but doens't activate it. User will have to press `Enter` to active it |
```
我覺得 Chakra 用的這個命名 `isManual` 沒有很直覺，也許可以改成 `activateOnKeypress` 之類的

另外這句應該改成 


If `true`, keyboard navigation changes focus between tabs but doens't activate it. User will have to press `Enter` or `Space` to activate it.

joshuatai replied:
對～～～Agree～～立馬改掉～～

joshuatai replied:
想不到其它名字 XD 就用 activateOnKeypress 吧！

tinaClin commented on packages/styled-ui-docs/pages/tabs.mdx:
```diff
@@ -0,0 +1,282 @@
+    <Tabs>
+      <TabList>
+        <Tab>One</Tab>
+        <Tab disabled>Two</Tab>
```
有個bug，在這個 example裡，即使該tab已經`disabled`了，但用鍵盤的左右鍵(← →)還是可以切換到disabled tab的TabPanel耶！

joshuatai replied:
嗯～～原本查克拉的 isDisabled 有一個地方沒改到

tinaClin commented on packages/styled-ui-docs/pages/tabs.mdx:
```diff
@@ -0,0 +1,282 @@
+By default, `Tabs` are activated automatically. This means when you use the
+arrow keys to change tabs, the tab is activated and focused.
+
+> The content of a `TabPanel` should ideally be preloaded. However, if switching
```
這樣子的縮排在畫面上，看起來有點怪~
![image](https://user-images.githubusercontent.com/24446505/87115148-08211680-c2a5-11ea-8b12-210a805ee862.png)

Commit:
fix: Fix disabled tab behavior and isManual props naming
